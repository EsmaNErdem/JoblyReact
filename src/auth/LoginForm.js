import React, {useState} from "react";
import { useHistory } from "react-router-dom";
import Alert from "../utilities/Alert";

/**
 * LoginForm Component
 * Displays form and renders controlled components functionality
 * 
 * A form component for user login. Handles user input for username, password.
 * Displays form errors if registration fails.
 * On submission:
 * - calls login function prop
 * - redirects to / route
 *
 * Routes ==> LoginForm ==> Alert
 * Routed as /login6
 * 
 */

const LoginForm = ({ login }) => {
    const history = useHistory();
    const INITIAL_STATE = {
        username: "",
        password: ""
    }
    const [formData, setFormData] = useState(INITIAL_STATE)
    const [formErrors, setFormErrors] = useState([]);

    console.debug(
        "LoginForm",
        "login=", typeof login,
        "formData=", formData,
        "formErrors=", formErrors,
    );
  
    /**
     * Handles form submission. Calls the login function with the form data.
     * Redirects to the home page upon successful registration, or displays form errors.
     */
    const handleSubmit = async e => {
        e.preventDefault();
        const result = await login(formData)
        if(result.success) {
            history.push("/")
        } else {
            setFormErrors(result.error)
        }
    }
    
    /** Update form data field */
    const handleChange = e => {
        const { name, value } = e.target
        setFormData(data => ({
            ...data,
            [name] : value
        }))
    }

    return (
        <div className="SignUpForm">
            <form onSubmit={handleSubmit}>
                <label htmlFor="username">Username</label>
                <input
                    id="username"
                    name="username"
                    value={formData.username}
                    onChange={handleChange}
                    placeholder="Username"
                />
                <label htmlFor="password">Password</label>
                <input
                    id="password"
                    name="password"
                    type="password"
                    value={formData.password}
                    onChange={handleChange}
                    placeholder="Password"
                />
                
                {formErrors.length
                    ? <Alert type="danger" messages={formErrors} />
                    : null
                }

                <button>Submit</button>
            </form>
        </div>
    )
}

export default LoginForm;