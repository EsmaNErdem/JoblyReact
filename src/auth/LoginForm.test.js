import { render, fireEvent } from "@testing-library/react";
import { MemoryRouter } from 'react-router-dom';
import LoginForm from "./LoginForm"

test("renders without crashing", async () => {
    render(<LoginForm />); 
});

test("it renders and matches with snaphot", () => {
    const { asFragment } = render(<LoginForm />);
    expect(asFragment()).toMatchSnapshot();
});

test("LoginForm renders correctly", () => {
    const { getByLabelText, getByText } = render(<LoginForm />);
    
    const usernameInput = getByLabelText("Username");
    const passwordInput = getByLabelText("Password");
    const submitButton = getByText("Submit");
  
    expect(usernameInput).toBeInTheDocument();
    expect(passwordInput).toBeInTheDocument();
    expect(submitButton).toBeInTheDocument();
  });
  
  test("Submitting the form calls login function", () => {
    const mockLogin = jest.fn().mockResolvedValue({ success: true }); // Mock login function
    const { getByLabelText, getByText } = render(
          <MemoryRouter >
              <LoginForm login={mockLogin} />
          </MemoryRouter>
      );
    
    const usernameInput = getByLabelText("Username");
    const passwordInput = getByLabelText("Password");
    const submitButton = getByText("Submit");
  
    fireEvent.change(usernameInput, { target: { value: "username" } });
    fireEvent.change(passwordInput, { target: { value: 123456789 } });
    fireEvent.click(submitButton);
  
    expect(mockLogin).toHaveBeenCalledWith({
      username: "username",
      password: "123456789"
    });
  });
   
