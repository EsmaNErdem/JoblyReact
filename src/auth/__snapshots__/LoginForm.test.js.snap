// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it renders and matches with snaphot 1`] = `
<DocumentFragment>
  <div
    class="SignUpForm"
  >
    <form>
      <label
        for="username"
      >
        Username
      </label>
      <input
        id="username"
        name="username"
        placeholder="Username"
        value=""
      />
      <label
        for="password"
      >
        Password
      </label>
      <input
        id="password"
        name="password"
        placeholder="Password"
        type="password"
        value=""
      />
      <button>
        Submit
      </button>
    </form>
  </div>
</DocumentFragment>
`;
